### PR TITLE
Add basic leader election

### PR DESCRIPTION
**Pull request by sergeyshevch**
Signed-off-by: Sergey Shevchenko <sergey.shevchenko@dteam.dev>

Added basic leader election for HA model

ref: #188 

TODO:
- [ ] Add documentation about HA mode
- [ ] Add HA deployment manifest
- [ ] Add HA option for helm chart
- [ ] Sqash all commits after reviews

@alexmt @ryota-sakamoto Can you look on this PR?

### CODE REVIEW

File: cmd/controller.go [hunk 1]
```diff
@@ -101,15 +117,37 @@ func newControllerCommand() *cobra.Command {
 			ctrl := controller.NewController(k8sClient, dynamicClient, argocdService, namespace, appLabelSelector, registry)
 			err = ctrl.Init(context.Background())
 			if err != nil {
-				return err
+				panic(err)
```
Comment (ryota-sakamoto) [inline cmd/controller.go:120]: `return err`

File: cmd/controller.go [hunk 2]
```diff
@@ -101,15 +117,37 @@ func newControllerCommand() *cobra.Command {
...
 			}
 
-			go ctrl.Run(context.Background(), processorsCount)
-			<-context.Background().Done()
 			return nil
 		},
 	}
 	clientConfig = k8s.AddK8SFlagsToCmd(&command)
+	options.BindLeaderElectionFlags(leaderElectionConfiguration, command.Flags())
+	command.Flags().StringVar(&leaderElectionId, "leader-election-id", uuid.New().String(), "Instance id for leader election.")
```
Comment (ryota-sakamoto) [inline cmd/controller.go:149]: I think it is suitable we use pod name instead of uuid

File: cmd/controller.go [hunk 3]
```diff
@@ -101,15 +117,37 @@ func newControllerCommand() *cobra.Command {
...
+					})
+				if err != nil {
+					return err
+				}
+
+				go leaderElector.Run(context.Background())
+				<-context.Background().Done()
+			} else {
+				go ctrl.Run(context.Background(), processorsCount)
+				<-context.Background().Done()
```
Comment (ryota-sakamoto) [inline cmd/controller.go:141]: I think we can make it common for line of 138 and 141.

**Review by ryota-sakamoto — CHANGES_REQUESTED** [review]
I think it makes no big problem, and I am waiting adding docs.

**Comment by sergeyshevch** [comment]
Thanks for review. I will add docs for it today or tomorrow

File: cmd/controller.go (hunk 2 above)
Comment (sergeyshevch) [reply cmd/controller.go:149]: It's an option. But i wasn't found clear solution how to get current pod. Also documentation for k8s leader election contain examples with uuid. I will look on it one more time

File: cmd/controller.go (hunk 3 above)
Comment (sergeyshevch) [reply cmd/controller.go:141]: I'm not sure how to do it. I guess i can't assign 137 or 140 line expression to some var. Can you suggest how to do it?

**Pull request closed without merging**